modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1245 @@
+      var instance = getPrivatizedThis;
+      forall segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        if segment.acquireIfNonEmpty(STATUS_REMOVE) {
+          var block = segment.headBlock;
+          while block != nil {
+            var tmp = block;
+            block = block.next;
+            delete tmp;
+          }
+
+          segment.headBlock = nil;
+          segment.tailBlock = nil;
+          segment.nElems.write(0);
+          segment.releaseStatus();
+        }
+      }
+    }
+
+    localThis.concurrentTasks.sub(1);
+  }
+
+  /*
+    Triggers a more static approach to load balancing; note that this operation
+    is parallel-safe but is performed in parallel across all `targetLocales` and
+    should only be called at a point of synchronization.
+  */
+  proc balance() {
+    var localThis = getPrivatizedThis;
+    if !BAG_NO_FREEZE {
+      local {
+        localThis.concurrentTasks.add(1);
+
+        // Check if the queue is now 'immutable'.
+        if localThis.frozenState.read() > FREEZE_UNFROZEN {
+          localThis.concurrentTasks.sub(1);
+          return;
+        }
+      }
+    }
+
+    // Phase 1: Acquire all locks from from first node and segment to last
+    // node and segment (our global locking order...)
+    for loc in localThis.targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      for segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        segment.acquire(STATUS_BALANCE);
+      }
+    }
+
+    // Phase 2: Concurrently redistribute elements from segments which contain
+    // more than the computed average.
+    coforall segmentIdx in 0 .. #here.maxTaskPar {
+      var nSegmentElems : [localThis.targetLocales.size] int;
+      var locIdx = 0;
+      for loc in localThis.targetLocales do on loc {
+        var nElems = getPrivatizedThis.bag.segments[segmentIdx].nElems.read() : int;
+        nSegmentElems[locIdx] = nElems;
+        locIdx += 1;
+      }
+
+      // Find the average and the excess. The excess is calculated as the amount
+      // of elements a segment has over the average, which is used to calculate
+      // the buffer size for each segment.
+      var total = (+ reduce nSegmentElems);
+      var avg = total / locIdx;
+      var excess : int;
+      for nElems in nSegmentElems {
+        if nElems > avg {
+          excess += nElems - avg;
+        }
+      }
+
+      // Allocate buffer, which holds the 'excess' elements for redistribution.
+      // Then fill it.
+      var buffer = c_malloc(eltType, excess);
+      var bufferOffset = 0;
+      for loc in localThis.targetLocales do on loc {
+        var average = avg;
+        ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+        var nElems = segment.nElems.read() : int;
+        if nElems > average {
+          var take = nElems - average;
+          var tmpBuffer = __primitive("+", buffer, bufferOffset);

```

Comment:
I don't believe so. I'm not sure how it is implemented, but since the buffer is allocated on another node, dereferencing 'buffer[bufferOffset]' at all would cause it to dereference invalid memory. Would it work because `buffer` is taken by reference and would call a `C_Ptr` method to handle this in a delicate enough manner?